File: python/0014.py
from leetcode_tester import Tester

from typing import Optional, List


# Vertical compare approach: compare each character from the first one and
# return immediate when
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        shortest = min(strs, key=len)

        for i, ch in enumerate(shortest):
            if not all([s[i] == ch for s in strs]):
                return shortest[:i]
        return shortest


if __name__ == '__main__':
    solution = Solution()
    test = Tester(solution.longestCommonPrefix)

    test.addTest(
        ["flower", "flow", "flight"], 'fl'
    )
    test.addTest(
        ["dog", "racecar", "car"], ''
    )
    test.addTest(
        [""], ''
    )
    test.addTest(
        ["a"], 'a'
    )
    test.doTest()

File: c++/0014.cpp
class Solution {
   public:
    string longestCommonPrefix(vector<string> &strs) {
        if (strs.size() == 1) {
            return strs[0];
        }

        sort(strs.begin(), strs.end());

        auto &start = strs.front();
        auto &end = strs.back();
        int n = min(start.size(), end.size());
        for (int i = 0; i < n; i++) {
            if (start[i] != end[i]) {
                return start.substr(0, i);
            }
        }
        return start.substr(0, n);
    }
};
